src/simulator/pysimtest/_pysimtest/runner.py

```diff
@@ -0,0 +1,216 @@
+        vehicle, scene and objects
+
+        This
+
+        :param test: the test to configure the simulation with
+        """
+        if (test.is_dynamic):
+            self._launch_dynamics()
+
+        scene_nodes     = self._configure_scene(test.scene)
+        vehicle_nodes   = self._configure_vehicle(test.vehicle)
+        object_nodes    = self._configure_objects(test.objects)
+        interface_nodes = self._configure_interfaces()
+
+        tree  = xmltree.parse(self._lib_path +
+                              "/uwsim/xml/in.xml")
+        root  = tree.getroot()
+        nodes = (scene_nodes + vehicle_nodes +
+                 object_nodes + interface_nodes)
+        for node in nodes:
+            # Insert node at end of tree to maintain UWsim doctype order
+            root.insert(len(list(root)),node)
+        header = ("<?xml version=\"1.0\" ?>\n" +
+                  "<!DOCTYPE UWSimScene SYSTEM \"UWSimScene.dtd\" >\n")
+        with open(self._lib_path + "/uwsim/xml/out.xml", 'wb') as file:
+            file.write(header)
+            tree.write(file)
+
+    def run_simulation(self, timeout):
+        """
+        Runs the uwsim simulation process
+        in a subprocess. The simulation is stopped
+        when the timeout length passes
+
+        :param timeout: the simulation timeout length
+        """
+        cmd = ["rosrun", "uwsim", "uwsim_binary", "--dataPath",
+                self._lib_path + "/uwsim/data", "--configfile",
+                self._lib_path + "/uwsim/xml/out.xml"]
+        proc = subprocess.Popen(cmd,
+                                stdout=constants.DEVNULL,
+                                stderr=constants.DEVNULL,
+                                preexec_fn=os.setsid)
+        self._uwsim_pgid = proc.pid
+        time.sleep(timeout)
+        self.stop_simulation()
+
+    def stop_simulation(self):
+        """
+        Stops the current execution of the simulation
+        if it is still running, alongside the dynamics
+        if they are running as well
+        """
+        if self._dynamics_pgid is not None:
+            os.killpg(os.getpgid(self._dynamics_pgid), signal.SIGTERM)
+            self._dynamics_pgid = None
+        if self._uwsim_pgid is not None:
+            os.killpg(os.getpgid(self._uwsim_pgid), signal.SIGTERM)
+            self._uwsim_pgid = None
+
+    def _launch_vehicle_ai(self):
+        """
+        Launches ros simulator ai in subprocess.
+        """
+        cmd = ["roslaunch", "simulator", "simulator_ai.launch"]
+        proc = subprocess.Popen(cmd,
+                                stdout=constants.DEVNULL,
+                                stderr=subprocess.PIPE,
+                                preexec_fn=os.setsid)
+        self._launch_pgid = proc.pid
+
+    def _launch_dynamics(self):
+        """
+        Launches dynamics in subprocess
+        """
+        #TODO: figure out if this can be a seperate node or
+        #      if you need to relaunch the ai
```

Comment:
`TODO` should be removed or associated with an issue (same goes for other `TODO`'s in this file)

Reply:
Sound good, I will remove them, was not sure of the conventions.

Also to answer the general questions:

- Eventually I would like it to be for automated testing but as you can imagine it is quite difficult to assert that a certain condition is met while running an instance of the simulation. Right now, the library is meant to be used as a way to easily parametrize an instance of the simulation, and a given test case simply ends after its defined timeout length. I hope sometime in the future to research more deeply into an effective way to automate testing but as the competition is approaching soon, it is not of vital importance.

- Yes I was unsure of what to do with them, I will do as you suggested

- For sure, I will follow this convention from now on.

- I sincerely do apologize!! This definitely got away on me and I should have done a PR with the basic outline of how it is meant to work before diving right in to the implementation. I will make sure not to do this again!

Reply:
Ah, totally makes sense. Also don't worry about it, this was actually pretty painless as far as large reviews go.